lib/lambda_ethereum_consensus/ssz_ex/decode.ex

```diff
@@ -0,0 +1,379 @@
+
+      true ->
+        case bit_vector do
+          # Padding bits are clear
+          <<_first::binary-size(num_bytes - 1), 0::size(8 - rem(size, 8) &&& 7),
+            _rest::bitstring>> ->
+            {:ok, BitVector.new(bit_vector, size)}
+
+          _else ->
+            {:error, "ExcessBits"}
+        end
+    end
+  end
+
+  defp decode_fixed_list(binary, inner_type, size) do
+    fixed_size = Utils.get_fixed_size(inner_type)
+
+    with {:ok, decoded_list} = result <- decode_fixed_collection(binary, fixed_size, inner_type),
+         :ok <- check_valid_list_size_after_decode(size, length(decoded_list)) do
+      result
+    end
+  end
+
+  defp decode_fixed_vector(binary, inner_type, size) do
+    fixed_size = Utils.get_fixed_size(inner_type)
+
+    with :ok <- check_valid_vector_size_prev_decode(fixed_size, size, binary),
+         {:ok, decoded_vector} = result <-
+           decode_fixed_collection(binary, fixed_size, inner_type),
+         :ok <- check_valid_vector_size_after_decode(size, length(decoded_vector)) do
+      result
+    end
+  end
+
+  defp check_valid_vector_size_prev_decode(fixed_size, size, binary)
+       when fixed_size * size == byte_size(binary),
+       do: :ok
+
+  defp check_valid_vector_size_prev_decode(_fixed_size, _size, _binary),
+    do: {:error, "Invalid vector size"}
+
+  defp check_valid_vector_size_after_decode(size, decoded_size)
+       when decoded_size == size and decoded_size > 0,
+       do: :ok
+
+  defp check_valid_vector_size_after_decode(_size, _decoded_size),
+    do: {:error, "invalid vector decoded size"}
+
+  defp check_valid_list_size_after_decode(size, decoded_size) when decoded_size <= size, do: :ok
+
+  defp check_valid_list_size_after_decode(_size, _decoded_size),
+    do: {:error, "invalid max_size of list"}
+
+  defp decode_variable_list(binary, _, _) when byte_size(binary) == 0 do
+    {:ok, []}
+  end
+
+  defp decode_variable_list(
+         <<first_offset::integer-32-little, _rest_bytes::bitstring>>,
+         _inner_type,
+         size
+       )
+       when div(first_offset, @bytes_per_length_offset) > size,
+       do: {:error, "invalid length list"}
+
+  defp decode_variable_list(binary, inner_type, _size) do
+    <<first_offset::integer-32-little, rest_bytes::bitstring>> = binary
+    num_elements = div(first_offset, @bytes_per_length_offset)
+
+    if Integer.mod(first_offset, @bytes_per_length_offset) != 0 ||
+         first_offset < @bytes_per_length_offset do
+      {:error, "InvalidListFixedBytesLen"}
+    else
+      with :ok <-
+             sanitize_offset(first_offset, nil, byte_size(binary), first_offset) do
+        decode_variable_list_elements(
+          num_elements,
+          rest_bytes,
+          inner_type,
+          first_offset,
+          binary,
+          first_offset,
+          []
+        )
+        |> Enum.reverse()
+        |> Utils.flatten_results()
+      end
+    end
+  end
+
+  defp decode_variable_list_elements(
```

Comment:
is this handling the case were `num_elements` is `0`?

Reply:
It's done in line [#149](https://github.com/lambdaclass/lambda_ethereum_consensus/pull/983/files#diff-85ee24efe82c8f3df9daa43de90926eb4e761f11d46545f0fcce389e03108d3eR149).